src/librbd/internal.cc

```diff
@@ -2838,6 +2858,66 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force,
       return -EINVAL;
     }
 
+    // is mirroring  enabled for the child?
+    cls::rbd::MirrorImage mirror_image_internal;
+    r = cls_client::mirror_image_get(&ictx->md_ctx, ictx->id, &mirror_image_internal);
+    if (r == -ENOENT) {
+      // mirroring is not enabled for this image
+      ldout(cct, 20) << "ignoring disable command: mirroring is not enabled "
+        "for this image" << dendl;
+      return 0;
+    }else if (r == -EOPNOTSUPP) {
+      ldout(cct, 5) << "mirroring not supported by OSD" << dendl;
+      return r;
+    } else if (r < 0) {
+      lderr(cct) << "cannot disable mirroring: " << cpp_strerror(r) << dendl;
+      return r;
+    }
+    mirror_image_internal.state = cls::rbd::MIRROR_IMAGE_STATE_DISABLING;
+    r = cls_client::mirror_image_set(&ictx->md_ctx, ictx->id, mirror_image_internal);
+    if (r < 0) {
+      lderr(cct) << "cannot disable mirroring: " << cpp_strerror(r) << dendl;
+      return r;
+    }
+
+    RWLock::RLocker l(ictx->snap_lock);
+    map<librados::snap_t, SnapInfo> snap_info = ictx->snap_info;
+    for (auto &info : snap_info) {
+      librbd::parent_spec parent_spec(ictx->md_ctx.get_id(), ictx->id, info.first);
+      set<pair<string, string> > names;
+
+      r = list_children_of_snap(ictx, parent_spec, names);
+      if (r < 0)
+        return r;
+      if(names.empty())
+        continue;
+
+      IoCtx ioctx;
+      Rados rados(ictx->md_ctx);
+      for (auto &pair : names) {
+        r = rados.ioctx_create(pair.first.c_str(), ioctx);
+        if(r < 0)
+          return r;
+
+        ImageCtx *image_ctx = new ImageCtx(pair.second, "", NULL, ioctx, true);
```

Comment:
No need to open image -- ```list_children_of_snap``` internally collects a list of image ids. I would modify that function to return a pool id, pool name, and image id tuple. The current ```list_children``` function can lookup the image names from the ids and return the pool name / image name pair. This function can directly use the pool id (which is safer than using pool name due to renames) and image id.